proxy/http/HttpTransact.cc

```diff
@@ -6757,27 +6757,28 @@ HttpTransact::handle_response_keep_alive_headers(State *s, HTTPVersion ver, HTTP
     // to the client to keep the connection alive.
     // Insert a Transfer-Encoding header in the response if necessary.
 
-    // check that the client is HTTP 1.1 and the conf allows chunking or the client
-    // protocol unchunks before returning to the user agent (i.e. is http/2)
-    if (s->client_info.http_version == HTTPVersion(1, 1) &&
-        (s->txn_conf->chunking_enabled == 1 ||
-         (s->state_machine->plugin_tag && (!strncmp(s->state_machine->plugin_tag, "http/2", 6)))) &&
-        // if we're not sending a body, don't set a chunked header regardless of server response
-        !is_response_body_precluded(s->hdr_info.client_response.status_get(), s->method) &&
-        // we do not need chunked encoding for internal error messages
-        // that are sent to the client if the server response is not valid.
-        (((s->source == SOURCE_HTTP_ORIGIN_SERVER || s->source == SOURCE_TRANSFORM) && s->hdr_info.server_response.valid() &&
-          // if we receive a 304, we will serve the client from the
-          // cache and thus do not need chunked encoding.
-          s->hdr_info.server_response.status_get() != HTTP_STATUS_NOT_MODIFIED &&
-          (s->current.server->transfer_encoding == HttpTransact::CHUNKED_ENCODING ||
-           // we can use chunked encoding if we cannot trust the content
-           // length (e.g. no Content-Length and Connection:close in HTTP/1.1 responses)
-           s->hdr_info.trust_response_cl == false)) ||
-         // handle serve from cache (read-while-write) case
-         (s->source == SOURCE_CACHE && s->hdr_info.trust_response_cl == false) ||
-         // any transform will potentially alter the content length. try chunking if possible
-         (s->source == SOURCE_TRANSFORM && s->hdr_info.trust_response_cl == false))) {
+    // check that the client protocol is HTTP/1.1 and the conf allows chunking or
+    // the client protocol doesn't support chunked transfer coding (i.e. HTTP/1.0, HTTP/2, and HTTP/3)
+    if ((s->state_machine->client_protocol && strncmp(s->state_machine->client_protocol, "http/2", 6) == 0) ||
```

Comment:
Sigh, all three versions are checked in different ways...

Reply:
It's one of we need to figure out. Currently we have `int32_t` (HTTP/1.1), `char *` (HTTP/2), and `std::string_view` (HTTP/3).

Reply:
I think this doesn't have to be solved on this PR and it might be a part of @a-canary 's refactoring, but we should probably check whether a client protocol supports chunked response (bool) but not its version.

Reply:
Agreed. I'll try `ProxyTransaction::is_chunked_encoding_supported()`. It's currently used for only POST request.